### PR TITLE
Update type-inference of annotated expressions

### PR DESCRIPTION
**Pull request by sjakobi**
… as standardized in https://github.com/dhall-lang/dhall-lang/pull/745.

### CODE REVIEW

File: dhall/src/Dhall/TypeCheck.hs [hunk 1]
```diff
@@ -937,86 +941,93 @@ infer typer = loop
...
-                        return (Nothing, eval values _T₁)
+            if Data.Set.null diffU
+                then return ()
+                else die (MissingHandler diffU)
 
-                    Nothing -> do
-                        case Dhall.Map.uncons yTs' of
-                            Nothing -> do
-                                die MissingMergeType
+            let match y _T' mU' =
```
Comment (sjakobi) [inline dhall/src/Dhall/TypeCheck.hs]: I'm not fully convinced about my naming scheme – better ideas are welcome! :)

File: dhall/src/Dhall/TypeCheck.hs [hunk 2]
```diff
@@ -937,86 +941,93 @@ infer typer = loop
...
+            case mT₀' of
+                Nothing -> return ()
 
-                                                return (_T₂' (fresh ctx x))
+                Just _T₀' -> do
+                    let _T₀'' = quote names _T₀'
 
-                                            Nothing -> do
-                                                let _T'' = quote names _T'
+                    _ <- loop ctx _T₀''
```
Comment (sjakobi) [inline dhall/src/Dhall/TypeCheck.hs]: The free variable check here could alternatively be included in `match` above. That might slightly improve type errors like this one:

```
$ echo 'merge { a = None Bool, x = None } (<a | x: Type>.x Bool)' | dhall:dhall type
Error: Handlers should have the same output type

…
- Bool
+ A@-1

1│ merge { a = None Bool, x = None } (<a | x: Type>.x Bool)
```

File: dhall/src/Dhall/TypeCheck.hs [hunk 3]
```diff
@@ -937,86 +941,93 @@ infer typer = loop
...
+                            Nothing -> return (Just (y₁, _T₁'))
 
-                        Just _T' -> do
-                            _T₃' <- do
-                                case mU of
-                                    Nothing -> do
-                                        return _T'
+                            Just (y₀, _T₀') -> do
+                                if Eval.conv values _T₀' _T₁'
+                                    then return (Just (y₀, _T₀'))
```
Comment (sjakobi) [inline dhall/src/Dhall/TypeCheck.hs]: Would this be better?

```suggestion
                                    then return myT₀'
```

Comment (sjakobi) [reply dhall/src/Dhall/TypeCheck.hs]: Yes – it seems to prevent an additional allocation!

File: dhall/src/Dhall/TypeCheck.hs [hunk 4]
```diff
@@ -937,86 +937,85 @@ infer typer = loop
...
+                            if Eval.conv values _A₀' _A₁'
+                                then do
+                                    let _T₁' = _T₀' (fresh ctx x)
 
-                                                Nothing -> do
-                                                    let _T'' = quote names _T'
+                                    let _T₁'' = quote names _T₁'
 
-                                                    die (HandlerNotAFunction y₀ _T'')
+                                    _ <- loop ctx _T₁''
```
Comment (sjakobi) [inline dhall/src/Dhall/TypeCheck.hs]: I think the free variable check here can be done more efficiently, but I haven't yet figured out how…

File: dhall/src/Dhall/TypeCheck.hs [hunk 5]
```diff
@@ -937,86 +939,89 @@ infer typer = loop
...
+                                    -- If x appears in _T₁'', it's index will be -1.
+                                    -- Any well-typed Var has a non-negative index,
+                                    -- so if we find a negative index, we know
+                                    -- something is wrong.
+                                    let badVar (Var (V _ n)) = n < 0
+                                        badVar _             = False
 
-                                return (Just y₀, _T₁')
+                                    if badVar _T₁'' || Lens.Family.anyOf Dhall.Core.subExpressions badVar _T₁''
+                                        then die (UnboundVariable x)
```
Comment (sjakobi) [inline dhall/src/Dhall/TypeCheck.hs]: Does my logic above make sense? I'm not sure how to do this more efficiently without writing a lot of code (to inspect `Val`s directly).

I would create a new `TypeMessage` for this failure condition if the logic seems sound.

File: dhall/src/Dhall/TypeCheck.hs [hunk 6]
```diff
@@ -937,86 +939,89 @@ infer typer = loop
...
-                                        return _T'
+                            die (HandlerNotAFunction y handler'')
 
-                                    Just _A₁' -> do
-                                        case Eval.toVHPi _T' of
-                                            Just (x, _A₀', _T₂') -> do
-                                                if Eval.conv values _A₀' _A₁'
-                                                    then do
-                                                        return ()
+            let matched =
```
Comment (Gabriel439) [inline dhall/src/Dhall/TypeCheck.hs]: If you do:

```haskell
matched <- sequence (Data.Map.intersectionWithKey match (Dhall.Map.toMap yTs') (Dhall.Map.toMap yUs'))
```

... then you won't need to handle the `Either`s in `checkMatched`

**Comment by sjakobi** [comment]
@Gabriel439 Can you check whether my error message makes sense?

Sample output for the `MergeHandlerFreeVar` test case:

    Error: Dependently typed handler

    Explanation: You can ❰merge❱ the alternatives of a union using a record with one
    handler per alternative, like this:

        ┌─────────────────────────────────────────────────────────────────┐
        │ let union    = < Left : Natural | Right : Bool >.Left 2         │
        │ let handlers = { Left = Natural/even, Right = λ(x : Bool) → x } │
        │ in  merge handlers union : Bool                                 │
        └─────────────────────────────────────────────────────────────────┘

    ... but the type of the ❰merge❱ may not depend on the union value.

    For example, the following expression is not valid:

        ┌────────────────────────────────────┐
        │ let handlers = { x = None }        │
        │ let union    = < x : Type >.x Bool │
        │ in  merge handlers union           │
        └────────────────────────────────────┘

    Your handler for the following alternative:

    ↳ x

    ... has type:

    ↳ ∀(A : Type) → Optional A

    ... and its output type depends on the value of the variable:

    ↳ A

    ────────────────────────────────────────────────────────────────────────────────

    1│ merge { x = None } (<x: Type>.x Bool)

File: dhall/src/Dhall/TypeCheck.hs [hunk 7]
```diff
@@ -3140,6 +3155,47 @@ prettyTypeMessage (HandlerInputTypeMismatch expr0 expr1 expr2) =
         txt1 = insert expr1
         txt2 = insert expr2
 
+prettyTypeMessage (DependentlyTypedHandler label handlerType variable) =
+    ErrorMessages {..}
+  where
+    short = "Dependently typed handler"
```
Comment (Gabriel439) [inline dhall/src/Dhall/TypeCheck.hs]:
```suggestion
    short = "Disallowed handler type"
```

File: dhall/src/Dhall/TypeCheck.hs [hunk 8]
```diff
@@ -3140,6 +3155,47 @@ prettyTypeMessage (HandlerInputTypeMismatch expr0 expr1 expr2) =
...
+        \                                                                                \n\
+        \                                                                                \n\
+        \    ┌─────────────────────────────────────────────────────────────────┐         \n\
+        \    │ let union    = < Left : Natural | Right : Bool >.Left 2         │         \n\
+        \    │ let handlers = { Left = Natural/even, Right = λ(x : Bool) → x } │         \n\
+        \    │ in  merge handlers union : Bool                                 │         \n\
+        \    └─────────────────────────────────────────────────────────────────┘         \n\
+        \                                                                                \n\
+        \                                                                                \n\
+        \... but the type of the ❰merge❱ may not depend on the union value.              \n\
```
Comment (Gabriel439) [inline dhall/src/Dhall/TypeCheck.hs]:
```suggestion
        \... but the output type of a handler may not depend on the input type.         \n\
```

File: dhall/src/Dhall/TypeCheck.hs [hunk 9]
```diff
@@ -3140,6 +3155,47 @@ prettyTypeMessage (HandlerInputTypeMismatch expr0 expr1 expr2) =
...
+        \                                                                                \n\
+        \For example, the following expression is " <> _NOT <> " valid:                  \n\
+        \                                                                                \n\
+        \                                                                                \n\
+        \    ┌────────────────────────────────────┐                                      \n\
+        \    │ let handlers = { x = None }        │                                      \n\
```
Comment (Gabriel439) [inline dhall/src/Dhall/TypeCheck.hs]: Perhaps add an `Invalid: …` in the right-hand margin explaining why it is not valid.  For example, you can say something along the lines of: "Invalid: The output type is ❰Optional a❱, which references the input type ❰a❱"

File: dhall/src/Dhall/TypeCheck.hs [hunk 10]
```diff
@@ -3140,6 +3155,47 @@ prettyTypeMessage (HandlerInputTypeMismatch expr0 expr1 expr2) =
...
+        \                                                                                \n\
+        \Your handler for the following alternative:                                     \n\
+        \                                                                                \n\
+        \" <> insert label <> "\n\
+        \                                                                                \n\
+        \... has type:                                                                   \n\
+        \                                                                                \n\
+        \" <> insert handlerType <> "\n\
+        \                                                                                \n\
+        \... and its output type depends on the value of the variable:                   \n\
```
Comment (Gabriel439) [inline dhall/src/Dhall/TypeCheck.hs]: I suggest also specifically including the output type here, too.  In other words:

```
... and the handler's output type:

↳ …

... references the handler's input type:

↳ …
```

File: dhall/src/Dhall/TypeCheck.hs [hunk 11]
```diff
@@ -937,86 +939,98 @@ infer typer = loop
...
 
-                                        Just (Just _) -> do
-                                            case Eval.toVHPi _T' of
-                                                Just (x, _A₀', _T₀') -> do
-                                                    return (_T₀' (fresh ctx x))
+                                    let _T₁'' = quote names _T₁'
 
-                                                Nothing -> do
-                                                    let _T'' = quote names _T'
+                                    -- x appearing in _T₁'' would indicate a dependently typed
```
Comment (sjakobi) [inline dhall/src/Dhall/TypeCheck.hs]: Thanks for the suggestions on the error message. Now I'm wondering whether I was mistaken to call the handler "dependently typed", and whether I should also update this comment.

Comment (Gabriel439) [reply dhall/src/Dhall/TypeCheck.hs]: Yeah, I would just call it "disallowed" rather than "dependently typed"

**Comment by sjakobi** [comment]
I've made some adjustments, following some of your suggestions. However I'm still referring to the handler's input _value_, as that really appears to be the problem.

Sample output, this time for

    merge { y = (\\(x : Natural) -> assert : x === x) } (< y : Natural>.y 1)

```
$ echo 'merge { y = (\\(x : Natural) -> assert : x === x) } (< y : Natural>.y 1)' | cabal run -- dhall:dhall --explain
Up to date


Error: Disallowed handler type

Explanation: You can ❰merge❱ the alternatives of a union using a record with one
handler per alternative, like this:                                             
                                                                                
                                                                                
    ┌─────────────────────────────────────────────────────────────────┐         
    │ let union    = < Left : Natural | Right : Bool >.Left 2         │         
    │ let handlers = { Left = Natural/even, Right = λ(x : Bool) → x } │         
    │ in  merge handlers union : Bool                                 │         
    └─────────────────────────────────────────────────────────────────┘         
                                                                                
                                                                                
... but the output type of a handler may not depend on the input value.         
                                                                                
For example, the following expression is not valid:                  
                                                                                
                                                                                
      Invalid: The output type is ❰Optional A❱, which references the input      
      value ❰A❱.                                                                
                  ⇩                                                             
    ┌──────────────────────────────────────────┐                                
    │ merge { x = None } (< x : Type >.x Bool) │                                
    └──────────────────────────────────────────┘                                
                                                                                
                                                                                
Your handler for the following alternative:                                     
                                                                                
↳ y
                                                                                
... has type:                                                                   
                                                                                
↳ ∀(x : Natural) → x ≡ x
                                                                                
... where the output type:                                                      
                                                                                
↳ x ≡ x
                                                                                
... references the handler's input value:                                                 
                                                                                
↳ x

────────────────────────────────────────────────────────────────────────────────

1│ merge { y = (\(x : Natural) -> assert : x === x) } (< y : Natural>.y 1)

(stdin):1:1
```

What do you think @Gabriel439?

**Comment by Gabriel439** [comment]
@sjakobi: Yeah, that looks great!  I agree that input value is the right term to use

**Pull request merged**